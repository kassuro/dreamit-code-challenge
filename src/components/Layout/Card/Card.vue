<template>
  <div
    class="flex-col rounded-md border p-3 shadow-md"
    :class="{
      [cardSize]: true,
      [backgroundColor[background]]: true,
      'lg:max-w-3xl': size !== 'full',
      'cursor-pointer': clickable,
      'hover:brightness-90': clickable,
    }"
    @click="clickHandler"
    @keyup.enter="clickHandler"
  >
    <div v-if="$slots.title" class="mb-3">
      <slot name="title" />
    </div>

    <div class="flex-grow">
      <slot />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue';

// There seems to be a problem with volar here which makes exporting the interface currently necessary.
// Otherwise we get errors from volar / vue-tsc in VS-Code.
// Volar core maintainer suggests putting props in script tag, which seems weird as we want to use script setup.
// https://github.com/johnsoncodehk/volar/issues/1232
export interface Props {
  size: 'sm' | 'md' | 'lg' | 'full';
  background?: 'yellow' | 'red' | 'blue' | 'white';
  clickable?: boolean;
  clickAction?: () => void;
}

const props = withDefaults(defineProps<Props>(), {
  background: 'white',
  clickable: false,
  link: false,
  clickAction: () => {},
});

const cardSize = computed(() => {
  if (props.size === 'sm') {
    return 'w-1/4';
  }
  if (props.size === 'md') {
    return 'w-1/2';
  }
  return 'w-full';
});

const backgroundColor = {
  yellow: 'bg-yellow-500',
  red: 'bg-red-500',
  blue: 'bg-blue-500',
  white: 'bg-white-500',
};

const clickHandler = () => {
  if (!props.clickable) {
    return;
  }

  props.clickAction();
};
</script>

<script lang="ts">
export default {
  name: 'TottoCard',
};
</script>
