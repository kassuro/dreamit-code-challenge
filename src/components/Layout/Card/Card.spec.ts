import { describe, expect, it } from 'vitest';
import Card from '@/components/Layout/Card/Card.vue';
import { shallowMount } from '@vue/test-utils';

describe('Card.vue', () => {
  it('renders', () => {
    const wrapper = shallowMount(Card, {
      props: {
        size: 'sm',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders with size md', () => {
    const wrapper = shallowMount(Card, {
      props: {
        size: 'md',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders with size lg', () => {
    const wrapper = shallowMount(Card, {
      props: {
        size: 'lg',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders with size full', () => {
    const wrapper = shallowMount(Card, {
      props: {
        size: 'full',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('has the correct bg class', () => {
    const wrapper = shallowMount(Card, {
      props: {
        size: 'sm',
        background: 'red',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('div').classes().includes('bg-red-500')).toBeTruthy();
  });
});
