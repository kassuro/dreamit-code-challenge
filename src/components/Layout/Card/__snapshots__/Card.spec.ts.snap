// Vitest Snapshot v1

exports[`Card.vue > has the correct bg class 1`] = `
<div
  class="flex-col rounded-md border p-3 shadow-md w-1/4 bg-red-500 lg:max-w-3xl"
>
  <!--v-if-->
  <div
    class="flex-grow"
  >
    
    
  </div>
</div>
`;

exports[`Card.vue > renders 1`] = `
<div
  class="flex-col rounded-md border p-3 shadow-md w-1/4 bg-white-500 lg:max-w-3xl"
>
  <!--v-if-->
  <div
    class="flex-grow"
  >
    
    
  </div>
</div>
`;

exports[`Card.vue > renders with size full 1`] = `
<div
  class="flex-col rounded-md border p-3 shadow-md w-full bg-white-500"
>
  <!--v-if-->
  <div
    class="flex-grow"
  >
    
    
  </div>
</div>
`;

exports[`Card.vue > renders with size lg 1`] = `
<div
  class="flex-col rounded-md border p-3 shadow-md w-full bg-white-500 lg:max-w-3xl"
>
  <!--v-if-->
  <div
    class="flex-grow"
  >
    
    
  </div>
</div>
`;

exports[`Card.vue > renders with size md 1`] = `
<div
  class="flex-col rounded-md border p-3 shadow-md w-1/2 bg-white-500 lg:max-w-3xl"
>
  <!--v-if-->
  <div
    class="flex-grow"
  >
    
    
  </div>
</div>
`;
